File: m02_ros2_communication/m02_ros2_with_py/m02_ros2_with_py/int64_pub.py
# !/usr/bin/env python3
import random

# ------------------ ROS2 Depedencies ----------------------------------------
import rclpy
from rclpy.node import Node

# -------------------- ROS2 Messages -----------------------------------------
from std_msgs.msg import Int64

# ------------------- Publisher Class Node ------------------------------
class MyInt64Publisher(Node):
    """
    Simple publisher of an integer (64 bit) in a specific time range.

    Attributes
    ---
    pub : Publisher object
        Object that will communicate Int64 nums by using the /num_int64 topic

    timer - Timer object
        Time counter that links a callback when the period of time is copleted

    Methods
    ---
        num_callback()
            Publish a random number when called
    """
    def __init__(self):
        """
        Constructor that initialize node, set publisher and timer.
        """
        # Initialization of node (name must be unique)
        super().__init__('int64_pub')
        
        # Instance a publisher for integer that uses the /num_int64 topic and 
        # queue size of 10.
        self.pub = self.create_publisher(Int64, 'num_int64', 10)

        # Instance a timer for publishing info at the given time
        time_sleep = 2
        self.timer = self.create_timer(time_sleep, self.num_callback)
    
    def num_callback(self):
        """
        Publish a random integer when the timer indicates it.
        """
        # Instance an integer (64 bits) object and generate a random number
        msg = Int64()
        msg.data = random.randint(1, 1000)

        # Publish the given message and log the process.
        self.pub.publish(msg)
        self.get_logger().info('Publishing integer: %d' % msg.data )


# -------------------------- Implementation -----------------------------------
def main(args=None):
    """
    Simple program to publish an integer with rclpy by implementing a publisher 
    class.
    """

    # Initialize ROS Client Library for Python
    rclpy.init(args=args)
    
    # Instance integer publisher
    my_pub = MyInt64Publisher()

    # Spin (act as a loop everytime a publishing is complete)
    rclpy.spin(my_pub)

    # It is a good practice to destroy the node and shutdown rclpy, but if it
    # forgotten, the garbage collector will take care of it (in the future)
    my_pub.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: m02_ros2_communication/m02_ros2_with_cpp/src/int64_pub.cpp
/**
 * File: int64_pub.cpp
 * 
 * Description: Simple publisher for integer (64 bits) implemented with the 
 * rclcpp implementation for ROS.
*/

// ------------------------- CPP standard Libraries --------------------------

#include <chrono>     // For flexible collection of types that trask time with
                      // varying degrees of precision.

#include <functional> // Function objects library, related with hash functions,
                      // those function with outputs that depends only on the
                      // input.

#include <memory>     // General utilities to manage dynamic memory

#include <string>     // Management and usage of strings.

#include <cstdlib>    // General purpose functions 

// ------------------------ rclcpp headers -----------------------------------

#include "rclcpp/rclcpp.hpp"        // Proper include for using rclcpp 

#include "std_msgs/msg/int64.hpp"   // Include of standard message for int64

// ------------------------ Implementations of namespaces --------------------
using namespace std::chrono_literals;


// ------------------------- Int64 Publisher Class ---------------------------

/**
 * Simple class for a publisher of integers (64 bits) which inherits from Node
 * class.
*/
class IntPub : public rclcpp::Node
{
  // Public interface
  public:
    /**
     * Constructor call that initialize the node name (using parent interface),
     * instance a publisher object for integers by using the 'num_int64'
     * topic and linking a callback with a timer.
    */
    IntPub() : Node("int64_pub"), count_(0)
    {
      publisher_ = this->create_publisher<std_msgs::msg::Int64>("num_int64", 10);
      timer_ = this->create_wall_timer(
      500ms, std::bind(&IntPub::timer_callback, this));
    }
  
  // Private interface
  private:
    /**
     * Method for publishing the integer when the established time for the
     * timer is completed.
    */
    void timer_callback()
    {
      // Defining lower and upper limit for random
      auto ll = 1, ul = 100;

      // Instance standard message obj for integers
      auto message = std_msgs::msg::Int64();

      // Updating content with random number
      message.data = rand() % (ul - ll + 1) + ll;

      // Log and publish info
      RCLCPP_INFO(this->get_logger(), "Publishing (%d): %d", count_ ,message.data);
      publisher_->publish(message);
    }

    // Private shared pointer for timer instance
    rclcpp::TimerBase::SharedPtr timer_;

    // Private shared pointer for publisher instance
    rclcpp::Publisher<std_msgs::msg::Int64>::SharedPtr publisher_;
    
    // Counter
    size_t count_;
};



int main(int argc, char * argv[])
{
  // ROS Client Library for C++ initialization
  rclcpp::init(argc, argv);

  // Spin the publisher 
  rclcpp::spin(std::make_shared<IntPub>());

  // Closing node
  rclcpp::shutdown();
  return 0;
}